src/video/odk2/odk2.cc

```diff
@@ -0,0 +1,191 @@
+                                          1.0 - (2.0 * ((qy * qy) + (qz * qz)))));
+        const uint32_t headingIdx = turn_t(heading + 180_deg).value() * BAND_WIDTH;
+
+        LOGD << "Processing frame heading=" << headingIdx;
+
+        // Loop through rows of horizontal band
+        for(int i = 0; i < BAND_HEIGHT; i++) {
+            // Get pointer to rows in outFrame and mask
+            // **NOTE** we want to flip output
+            cv::Vec3b *outPixel = outFrame.ptr<cv::Vec3b>(BAND_HEIGHT - i - 1);
+            uint8_t *outMask = m_Mask.ptr<uint8_t>(BAND_HEIGHT - i - 1);
+
+            // Loop through columns of horizontal band
+            for (int j = 0; j < BAND_WIDTH; j++) {
+                const int bandIdx = (i * BAND_WIDTH) + ((j + headingIdx) % BAND_WIDTH);
+
+                // Get fourpi index to map to band pixel
+                const int fourpiIdx = band_to_fourpi(bandIdx, bandType_horz);
+
+                // Create pixel into OpenCV image
+                const cv::Vec3b pixel(m_RawFrameBuf.image[fourpiIdx * 3],
+                                      m_RawFrameBuf.image[(fourpiIdx * 3) + 1],
+                                      m_RawFrameBuf.image[(fourpiIdx * 3) + 2]);
+
+                // Write pixel to OpenCV image
+                *outPixel++ = pixel;
+
+                // If pixel is black, clear mask otherwise set
+                *outMask++ = (pixel[0] == 0 && pixel[1] == 0 && pixel[2] == 0) ? 0 : 255;
+            }
+        }
+
+        return true;
+    }
+    else {
+        return false;
+    }
+}
+//------------------------------------------------------------------------
+std::array<degree_t, 3> ODK2::getEulerAngles() const
+{
+    // Extract quaternion components from frame
+    m_OutputBufsMutex.lock();
+    const float qw = m_IMUBuf.q.w;
+    const float qx = m_IMUBuf.q.x;
+    const float qy = m_IMUBuf.q.y;
+    const float qz = m_IMUBuf.q.z;
+    m_OutputBufsMutex.unlock();
+
+    // **NOTE** this whole process should be doable using Eigen but I can't make it work so
+    // used https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+    // Calculate heading from quaternion
+    const radian_t heading(std::atan2(2.0 * ((qw * qz) + (qx * qy)),
+                                      1.0 - (2.0 * ((qy * qy) + (qz * qz)))));
+
+    // Calculate roll from quaternion
+    const radian_t roll(std::atan2(2.0 * ((qw * qx) + (qy * qz)),
+                                   1.0 - (2.0 * ((qx * qx) + (qy * qy)))));
+
+    // Calculate pitch from quaternion
+    const double sinp = 2.0 * ((qw * qy) - (qz * qx));
+    const radian_t pitch((std::abs(sinp) >= 1.0) ? std::copysign(M_PI / 2.0, sinp) : std::asin(sinp));
+
+    // To match BN055, we want to return in heading, roll, pitch
+    return {heading, roll, pitch};
+}
+//------------------------------------------------------------------------
+void ODK2::readThread()
```

Comment:
If a wire's loose you'll get a ``DEVKIT_DRIVER_RX_TIMEOUT`` which just logs a warning so, as BoB robotics generally uses exceptions for error handling I'll probably do that here. I feel it would be kinda nicer if the exceptions were returned on the main thread via ``readFrame``/``getEulerAngles`` rather than at a global level but that would be inconsistent with the rest of the library so I'll use the ``BackgroundExceptionCatcher``